runtime/perfetto/db.go

```diff
@@ -155,47 +143,269 @@ func (d *DB) Close() error {
 	if err != nil {
-		return fmt.Errorf("write trace to %s: %w", d.fname, err)
+		return nil, err
 	}
-	return nil
+	defer rows.Close()
+	var traces []TraceSummary
+	for rows.Next() {
+		if err := rows.Err(); err != nil {
+			return nil, err
+		}
+		var traceID string
+		var startTimeUs, endTimeUs int64
+		if err := rows.Scan(&traceID, &startTimeUs, &endTimeUs); err != nil {
+			return nil, err
+		}
+		traces = append(traces, TraceSummary{
+			TraceID:   traceID,
+			StartTime: time.UnixMicro(startTimeUs),
+			EndTime:   time.UnixMicro(endTimeUs),
+		})
+	}
+	return traces, nil
 }
 
-func (d *DB) encodeSpans(ctx context.Context, app, version string, spans []sdktrace.ReadOnlySpan) ([]byte, error) {
-	var b bytes.Buffer
+func (d *DB) fetchSpans(ctx context.Context, traceID string) ([]*protos.Span, error) {
+	const query = `SELECT encoded_span FROM spans WHERE trace_id=?`
+	rows, err := d.queryDB(ctx, query, traceID)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	var spans []*protos.Span
+	for rows.Next() {
+		if err := rows.Err(); err != nil {
+			return nil, err
+		}
+		var encoded []byte
+		if err := rows.Scan(&encoded); err != nil {
+			return nil, err
+		}
+		span := &protos.Span{}
+		if err := proto.Unmarshal(encoded, span); err != nil {
+			return nil, err
+		}
+		spans = append(spans, span)
+	}
+	return spans, nil
+}
+
+func (d *DB) queryDB(ctx context.Context, query string, args ...any) (*sql.Rows, error) {
+	// Keep retrying as long as we are getting the "locked" error.
+	for r := retry.Begin(); r.Continue(ctx); {
+		rows, err := d.db.QueryContext(ctx, query, args...)
+		if isLocked(err) {
+			continue
+		}
+		return rows, err
+	}
+	return nil, ctx.Err()
+}
+
+func (d *DB) execDB(ctx context.Context, query string, args ...any) (sql.Result, error) {
+	// Keep retrying as long as we are getting the "locked" error.
+	for r := retry.Begin(); r.Continue(ctx); {
+		res, err := d.db.ExecContext(ctx, query, args...)
+		if isLocked(err) {
+			continue
+		}
+		return res, err
+	}
+	return nil, ctx.Err()
+}
+
+// Serve runs an HTTP server that serves traces stored in the database.
+// In order to view the traces, open the following URL in a Chrome browser:
+//
+//	https://ui.perfetto.dev/#!/?url=http://<hostname>:9001?trace_id=<trace_id>
+//
+// , where <hostname> is the hostname of the machine running this server
+// (e.g., "127.0.0.1"), and <trace_id> is the trace ID in hexadecimal format.
+//
+// Perfetto UI requires that the server runs on the local port 9001. For that
+// reason, this method will block until port 9001 becomes available.
+func (d *DB) Serve(ctx context.Context) {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/status", func(w http.ResponseWriter, _ *http.Request) {
+		w.Write([]byte("OK")) //nolint:errcheck // response write error
+	})
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
```

Comment:
This handler is pretty big. Consider extracting it out into a top-level DB method:
    mux.HandleFunc("/", d.fetchTrace)